manager/controlapi/extension.go

```diff
@@ -22,13 +47,73 @@ func (s *Server) CreateExtension(ctx context.Context, request *api.CreateExtensi
 // - Returns `InvalidArgument` if the `GetExtensionRequest.extension_id` is empty.
 // - Returns an error if the get fails.
 func (s *Server) GetExtension(ctx context.Context, request *api.GetExtensionRequest) (*api.GetExtensionResponse, error) {
-	return nil, errors.New("not implemented")
+	if request.ExtensionID == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "extension ID must be provided")
+	}
+
+	var extension *api.Extension
+	s.store.View(func(tx store.ReadTx) {
+		extension = store.GetExtension(tx, request.ExtensionID)
+	})
+
+	if extension == nil {
+		return nil, status.Errorf(codes.NotFound, "extension %s not found", request.ExtensionID)
+	}
+
+	return &api.GetExtensionResponse{Extension: extension}, nil
 }
 
 // RemoveExtension removes the extension referenced by `RemoveExtensionRequest.ID`.
 // - Returns `InvalidArgument` if `RemoveExtensionRequest.extension_id` is empty.
 // - Returns `NotFound` if the an extension named `RemoveExtensionRequest.extension_id` is not found.
 // - Returns an error if the deletion fails.
 func (s *Server) RemoveExtension(ctx context.Context, request *api.RemoveExtensionRequest) (*api.RemoveExtensionResponse, error) {
-	return nil, errors.New("not implemented")
+	if request.ExtensionID == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "extension ID must be provided")
+	}
+
+	err := s.store.Update(func(tx store.Tx) error {
+		// Check if the extension exists
+		extension := store.GetExtension(tx, request.ExtensionID)
+		if extension == nil {
+			return status.Errorf(codes.NotFound, "could not find extension %s", request.ExtensionID)
+		}
+
+		// Check if any resources of this type present in the store, return error if so
+		resources, err := store.FindResources(tx, store.ByKind(request.ExtensionID))
+		if err != nil {
+			return status.Errorf(codes.Internal, "could not find resources using extension %s: %v", request.ExtensionID, err)
+		}
+
+		if len(resources) != 0 {
+			resourceNames := make([]string, 0, len(resources))
+			for _, resource := range resources {
+				resourceNames = append(resourceNames, resource.Annotations.Name)
+			}
+
+			extensionName := extension.Annotations.Name
+			resourceNameStr := strings.Join(resourceNames, ", ")
+			resourceStr := "resources"
+			if len(resourceNames) == 1 {
+				resourceStr = "resource"
+			}
+
+			return status.Errorf(codes.InvalidArgument, "extension '%s' is in use by the following %s: %v", extensionName, resourceStr, resourceNameStr)
```

Comment:
maybe stick to including the first few resources, and not all of them. if the user tries to delete an extension with thousands of outstanding resources, it's possible that the resulting error string may be too large to include in the response, and the user would get something like `message size exceeded`.